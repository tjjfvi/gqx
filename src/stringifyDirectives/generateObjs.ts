
import { Obj } from ".";
import { Context } from "..";

export default (ctx: Context) =>
  Object.entries(ctx.objectTypes).map(([type, fields]): Obj => ({
    type,
    props: fields.map(f => ({
      type,
      prop: f.name.value,
      dirs: (f.directives || []).map(d => ({
        type,
        prop: f.name.value,
        dirName: d.name.value,
        dirArgs: d.arguments?.slice(),
      }))
    })),
  }))
