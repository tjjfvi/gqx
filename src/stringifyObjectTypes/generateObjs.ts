
import { Context, Obj, Id, Prop } from ".";
import unwrapType from "../unwrapType";

export default (ctx: Context): [Obj[], Id[], Context] => {
  const ids: Id[] = [];
  const objs: Obj[] =
    Object.keys(ctx.objectTypes)
      .sort()
      .map(name => {
        const { fields, unions, implements: impls } = ctx.objectTypes[name];
        const props = fields.map(field => {
          const id: Id = {
            type: name,
            prop: field.name.value,
          };
          ids.push(id);
          const [{ name: { value: typeName } }, wrap,, wrapHKT] = unwrapType(field.type);
          const shallow = !(typeName in ctx.objectTypes);
          let prop: Prop = {
            id,
            type: typeName,
            shallow,
            wrap,
            wrapHKT,
            args: field.arguments?.slice() ?? [],
            directives: field.directives?.slice() ?? [],
            loc: field.loc,
          };
          return prop;
        }).sort((p, q) => p.id.prop > q.id.prop ? 1 : -1);
        const shallowProps = props.filter(p => p.shallow);
        const deepProps = props.filter(p => !p.shallow);
        const obj: Obj = {
          type: name,
          shallowProps,
          deepProps,
          props,
          unions,
          implements: impls,
          isBase: ctx.baseTypes.includes(name),
        };
        return obj;
      });
  return [objs, ids, ctx];
}
