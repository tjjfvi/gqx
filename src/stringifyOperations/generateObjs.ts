
import { Context, Obj, Id, Prop } from ".";
import unwrapType from "../unwrapType";

export default (ctx: Context): [Obj[], Context] => {
  const objs: Obj[] =
    ctx.operations
      .map(([prop, name]) => {
        const { fields } = ctx.objectTypes[name];
        const props = fields.map(field => {
          const id: Id = {
            type: name,
            prop: field.name.value,
          };
          const [{ name: { value: typeName } }, wrap] = unwrapType(field.type);
          let prop: Prop = {
            id,
            type: typeName,
            wrap,
            args: field.arguments?.slice() ?? [],
          };
          return prop;
        }).sort((p, q) => p.id.prop > q.id.prop ? 1 : -1);
        const obj: Obj = {
          type: name,
          prop,
          props,
        };
        return obj;
      })
  return [objs, ctx];
}
