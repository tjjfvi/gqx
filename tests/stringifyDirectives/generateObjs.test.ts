import generateObjs from "../../src/stringifyDirectives/generateObjs";

test("", () => {
  const ctx = utils.getOperations(["stringifyDirectives"]);
  let val = generateObjs(ctx);
  expect(val.map(v => [v.type, v.props.length, v.props.map(p => [p.prop, p.dirs.length])])).toMatchSnapshot();
  expect(val).toMatchSnapshot();
})
