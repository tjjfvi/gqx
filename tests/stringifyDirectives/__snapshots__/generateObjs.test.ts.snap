// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Array [
  Array [
    "Query",
    5,
    Array [
      Array [
        "noDirectives",
        0,
      ],
      Array [
        "oneDirective",
        1,
      ],
      Array [
        "multDirectivesArgs",
        4,
      ],
      Array [
        "multDirectivesNullArgs",
        4,
      ],
      Array [
        "multDirectivesNoArgs",
        5,
      ],
    ],
  ],
]
`;

exports[` 2`] = `
Array [
  Object {
    "props": Array [
      Object {
        "dirs": Array [],
        "prop": "noDirectives",
        "type": "Query",
      },
      Object {
        "dirs": Array [
          Object {
            "dirArgs": Array [
              Object {
                "kind": "Argument",
                "loc": Object {
                  "end": 351,
                  "start": 340,
                },
                "name": Object {
                  "kind": "Name",
                  "loc": Object {
                    "end": 341,
                    "start": 340,
                  },
                  "value": "x",
                },
                "value": Object {
                  "block": false,
                  "kind": "StringValue",
                  "loc": Object {
                    "end": 351,
                    "start": 343,
                  },
                  "value": "string",
                },
              },
            ],
            "dirName": "scalarArg",
            "prop": "oneDirective",
            "type": "Query",
          },
        ],
        "prop": "oneDirective",
        "type": "Query",
      },
      Object {
        "dirs": Array [
          Object {
            "dirArgs": Array [
              Object {
                "kind": "Argument",
                "loc": Object {
                  "end": 404,
                  "start": 395,
                },
                "name": Object {
                  "kind": "Name",
                  "loc": Object {
                    "end": 396,
                    "start": 395,
                  },
                  "value": "a",
                },
                "value": Object {
                  "block": false,
                  "kind": "StringValue",
                  "loc": Object {
                    "end": 404,
                    "start": 398,
                  },
                  "value": "test",
                },
              },
              Object {
                "kind": "Argument",
                "loc": Object {
                  "end": 414,
                  "start": 406,
                },
                "name": Object {
                  "kind": "Name",
                  "loc": Object {
                    "end": 407,
                    "start": 406,
                  },
                  "value": "b",
                },
                "value": Object {
                  "kind": "IntValue",
                  "loc": Object {
                    "end": 414,
                    "start": 409,
                  },
                  "value": "12345",
                },
              },
            ],
            "dirName": "scalarArg",
            "prop": "multDirectivesArgs",
            "type": "Query",
          },
          Object {
            "dirArgs": Array [
              Object {
                "kind": "Argument",
                "loc": Object {
                  "end": 464,
                  "start": 428,
                },
                "name": Object {
                  "kind": "Name",
                  "loc": Object {
                    "end": 429,
                    "start": 428,
                  },
                  "value": "a",
                },
                "value": Object {
                  "fields": Array [
                    Object {
                      "kind": "ObjectField",
                      "loc": Object {
                        "end": 437,
                        "start": 433,
                      },
                      "name": Object {
                        "kind": "Name",
                        "loc": Object {
                          "end": 434,
                          "start": 433,
                        },
                        "value": "x",
                      },
                      "value": Object {
                        "kind": "IntValue",
                        "loc": Object {
                          "end": 437,
                          "start": 436,
                        },
                        "value": "5",
                      },
                    },
                    Object {
                      "kind": "ObjectField",
                      "loc": Object {
                        "end": 449,
                        "start": 439,
                      },
                      "name": Object {
                        "kind": "Name",
                        "loc": Object {
                          "end": 440,
                          "start": 439,
                        },
                        "value": "y",
                      },
                      "value": Object {
                        "kind": "ListValue",
                        "loc": Object {
                          "end": 449,
                          "start": 442,
                        },
                        "values": Array [
                          Object {
                            "kind": "IntValue",
                            "loc": Object {
                              "end": 444,
                              "start": 443,
                            },
                            "value": "1",
                          },
                          Object {
                            "kind": "IntValue",
                            "loc": Object {
                              "end": 446,
                              "start": 445,
                            },
                            "value": "2",
                          },
                          Object {
                            "kind": "IntValue",
                            "loc": Object {
                              "end": 448,
                              "start": 447,
                            },
                            "value": "3",
                          },
                        ],
                      },
                    },
                    Object {
                      "kind": "ObjectField",
                      "loc": Object {
                        "end": 462,
                        "start": 451,
                      },
                      "name": Object {
                        "kind": "Name",
                        "loc": Object {
                          "end": 452,
                          "start": 451,
                        },
                        "value": "z",
                      },
                      "value": Object {
                        "fields": Array [
                          Object {
                            "kind": "ObjectField",
                            "loc": Object {
                              "end": 460,
                              "start": 456,
                            },
                            "name": Object {
                              "kind": "Name",
                              "loc": Object {
                                "end": 457,
                                "start": 456,
                              },
                              "value": "x",
                            },
                            "value": Object {
                              "kind": "IntValue",
                              "loc": Object {
                                "end": 460,
                                "start": 459,
                              },
                              "value": "5",
                            },
                          },
                        ],
                        "kind": "ObjectValue",
                        "loc": Object {
                          "end": 462,
                          "start": 454,
                        },
                      },
                    },
                  ],
                  "kind": "ObjectValue",
                  "loc": Object {
                    "end": 464,
                    "start": 431,
                  },
                },
              },
            ],
            "dirName": "inputArg",
            "prop": "multDirectivesArgs",
            "type": "Query",
          },
          Object {
            "dirArgs": Array [
              Object {
                "kind": "Argument",
                "loc": Object {
                  "end": 503,
                  "start": 482,
                },
                "name": Object {
                  "kind": "Name",
                  "loc": Object {
                    "end": 483,
                    "start": 482,
                  },
                  "value": "a",
                },
                "value": Object {
                  "kind": "ListValue",
                  "loc": Object {
                    "end": 503,
                    "start": 485,
                  },
                  "values": Array [
                    Object {
                      "block": false,
                      "kind": "StringValue",
                      "loc": Object {
                        "end": 489,
                        "start": 486,
                      },
                      "value": "a",
                    },
                    Object {
                      "block": false,
                      "kind": "StringValue",
                      "loc": Object {
                        "end": 493,
                        "start": 490,
                      },
                      "value": "b",
                    },
                    Object {
                      "block": false,
                      "kind": "StringValue",
                      "loc": Object {
                        "end": 497,
                        "start": 494,
                      },
                      "value": "c",
                    },
                    Object {
                      "kind": "NullValue",
                      "loc": Object {
                        "end": 502,
                        "start": 498,
                      },
                    },
                  ],
                },
              },
            ],
            "dirName": "scalarArrArg",
            "prop": "multDirectivesArgs",
            "type": "Query",
          },
          Object {
            "dirArgs": Array [
              Object {
                "kind": "Argument",
                "loc": Object {
                  "end": 548,
                  "start": 520,
                },
                "name": Object {
                  "kind": "Name",
                  "loc": Object {
                    "end": 521,
                    "start": 520,
                  },
                  "value": "a",
                },
                "value": Object {
                  "kind": "ListValue",
                  "loc": Object {
                    "end": 548,
                    "start": 523,
                  },
                  "values": Array [
                    Object {
                      "fields": Array [
                        Object {
                          "kind": "ObjectField",
                          "loc": Object {
                            "end": 528,
                            "start": 525,
                          },
                          "name": Object {
                            "kind": "Name",
                            "loc": Object {
                              "end": 526,
                              "start": 525,
                            },
                            "value": "x",
                          },
                          "value": Object {
                            "kind": "IntValue",
                            "loc": Object {
                              "end": 528,
                              "start": 527,
                            },
                            "value": "5",
                          },
                        },
                      ],
                      "kind": "ObjectValue",
                      "loc": Object {
                        "end": 529,
                        "start": 524,
                      },
                    },
                    Object {
                      "fields": Array [
                        Object {
                          "kind": "ObjectField",
                          "loc": Object {
                            "end": 534,
                            "start": 531,
                          },
                          "name": Object {
                            "kind": "Name",
                            "loc": Object {
                              "end": 532,
                              "start": 531,
                            },
                            "value": "x",
                          },
                          "value": Object {
                            "kind": "IntValue",
                            "loc": Object {
                              "end": 534,
                              "start": 533,
                            },
                            "value": "6",
                          },
                        },
                      ],
                      "kind": "ObjectValue",
                      "loc": Object {
                        "end": 535,
                        "start": 530,
                      },
                    },
                    Object {
                      "fields": Array [
                        Object {
                          "kind": "ObjectField",
                          "loc": Object {
                            "end": 546,
                            "start": 537,
                          },
                          "name": Object {
                            "kind": "Name",
                            "loc": Object {
                              "end": 538,
                              "start": 537,
                            },
                            "value": "y",
                          },
                          "value": Object {
                            "kind": "ListValue",
                            "loc": Object {
                              "end": 546,
                              "start": 539,
                            },
                            "values": Array [
                              Object {
                                "kind": "IntValue",
                                "loc": Object {
                                  "end": 541,
                                  "start": 540,
                                },
                                "value": "1",
                              },
                              Object {
                                "kind": "IntValue",
                                "loc": Object {
                                  "end": 543,
                                  "start": 542,
                                },
                                "value": "2",
                              },
                              Object {
                                "kind": "IntValue",
                                "loc": Object {
                                  "end": 545,
                                  "start": 544,
                                },
                                "value": "3",
                              },
                            ],
                          },
                        },
                      ],
                      "kind": "ObjectValue",
                      "loc": Object {
                        "end": 547,
                        "start": 536,
                      },
                    },
                  ],
                },
              },
            ],
            "dirName": "inputArrArg",
            "prop": "multDirectivesArgs",
            "type": "Query",
          },
        ],
        "prop": "multDirectivesArgs",
        "type": "Query",
      },
      Object {
        "dirs": Array [
          Object {
            "dirArgs": Array [
              Object {
                "kind": "Argument",
                "loc": Object {
                  "end": 603,
                  "start": 596,
                },
                "name": Object {
                  "kind": "Name",
                  "loc": Object {
                    "end": 597,
                    "start": 596,
                  },
                  "value": "a",
                },
                "value": Object {
                  "kind": "NullValue",
                  "loc": Object {
                    "end": 603,
                    "start": 599,
                  },
                },
              },
              Object {
                "kind": "Argument",
                "loc": Object {
                  "end": 612,
                  "start": 605,
                },
                "name": Object {
                  "kind": "Name",
                  "loc": Object {
                    "end": 606,
                    "start": 605,
                  },
                  "value": "b",
                },
                "value": Object {
                  "kind": "NullValue",
                  "loc": Object {
                    "end": 612,
                    "start": 608,
                  },
                },
              },
            ],
            "dirName": "scalarArg",
            "prop": "multDirectivesNullArgs",
            "type": "Query",
          },
          Object {
            "dirArgs": Array [
              Object {
                "kind": "Argument",
                "loc": Object {
                  "end": 633,
                  "start": 626,
                },
                "name": Object {
                  "kind": "Name",
                  "loc": Object {
                    "end": 627,
                    "start": 626,
                  },
                  "value": "a",
                },
                "value": Object {
                  "kind": "NullValue",
                  "loc": Object {
                    "end": 633,
                    "start": 629,
                  },
                },
              },
            ],
            "dirName": "inputArg",
            "prop": "multDirectivesNullArgs",
            "type": "Query",
          },
          Object {
            "dirArgs": Array [
              Object {
                "kind": "Argument",
                "loc": Object {
                  "end": 658,
                  "start": 651,
                },
                "name": Object {
                  "kind": "Name",
                  "loc": Object {
                    "end": 652,
                    "start": 651,
                  },
                  "value": "a",
                },
                "value": Object {
                  "kind": "NullValue",
                  "loc": Object {
                    "end": 658,
                    "start": 654,
                  },
                },
              },
            ],
            "dirName": "scalarArrArg",
            "prop": "multDirectivesNullArgs",
            "type": "Query",
          },
          Object {
            "dirArgs": Array [
              Object {
                "kind": "Argument",
                "loc": Object {
                  "end": 682,
                  "start": 675,
                },
                "name": Object {
                  "kind": "Name",
                  "loc": Object {
                    "end": 676,
                    "start": 675,
                  },
                  "value": "a",
                },
                "value": Object {
                  "kind": "NullValue",
                  "loc": Object {
                    "end": 682,
                    "start": 678,
                  },
                },
              },
            ],
            "dirName": "inputArrArg",
            "prop": "multDirectivesNullArgs",
            "type": "Query",
          },
        ],
        "prop": "multDirectivesNullArgs",
        "type": "Query",
      },
      Object {
        "dirs": Array [
          Object {
            "dirArgs": Array [],
            "dirName": "noArgs",
            "prop": "multDirectivesNoArgs",
            "type": "Query",
          },
          Object {
            "dirArgs": Array [],
            "dirName": "scalarArg",
            "prop": "multDirectivesNoArgs",
            "type": "Query",
          },
          Object {
            "dirArgs": Array [],
            "dirName": "inputArg",
            "prop": "multDirectivesNoArgs",
            "type": "Query",
          },
          Object {
            "dirArgs": Array [],
            "dirName": "scalarArrArg",
            "prop": "multDirectivesNoArgs",
            "type": "Query",
          },
          Object {
            "dirArgs": Array [],
            "dirName": "inputArrArg",
            "prop": "multDirectivesNoArgs",
            "type": "Query",
          },
        ],
        "prop": "multDirectivesNoArgs",
        "type": "Query",
      },
    ],
    "type": "Query",
  },
]
`;
